crates/ark/src/lsp/diagnostics.rs

```diff
@@ -471,6 +470,131 @@ fn recurse_assignment(
     ().ok()
 }
 
+fn handle_assignment_variable(
+    identifier: &Node,
+    context: &mut DiagnosticContext,
+) -> anyhow::Result<()> {
+    if !identifier.is_identifier_or_string() {
+        return Ok(());
+    }
+
+    let name = context.contents.node_slice(&identifier)?.to_string();
+    let range = identifier.range();
+    context.add_defined_variable(name.as_str(), range);
+
+    Ok(())
+}
+
+/// Support for dotty assignment
+///
+/// Comes in a few forms
+/// - `.[x, y] <- list(1, 2)`
+/// - `.[one, .., five] <- list(1, 2, 3, 4, 5)`
+/// - `.[foo = y] <- list(x = 1, y = 2, z = 3)`
+/// - `.[x, .[y, .[z]]] <- list(1, list(2, list(3)))`
+///
+/// https://cran.r-project.org/web/packages/dotty/index.html
+fn handle_assignment_dotty(
+    identifier: &Node,
+    context: &mut DiagnosticContext,
+) -> anyhow::Result<()> {
+    if !identifier.is_subset() {
+        // Not a subset call of the form `.[]`
+        return Ok(());
+    };
+
+    let Some(dot) = identifier.child_by_field_name("function") else {
+        return Ok(());
+    };
+    if !dot.is_identifier() {
+        return Ok(());
+    };
+
+    let dot = context.contents.node_slice(&dot)?;
+    if dot != "." {
+        return Ok(());
+    };
+
+    // Make sure we're not being invoked within a magrittr pipe, since
+    // '.' has special semantics in that scope.
+    if node_is_within_magrittr_pipe(identifier, context)? {
+        return Ok(());
+    }
+
+    // Iterate over each argument, and look for identifiers.
+    let Some(arguments) = identifier.child_by_field_name("arguments") else {
+        return Ok(());
+    };
+
+    let mut cursor = arguments.walk();
+
+    for child in arguments.children_by_field_name("argument", &mut cursor) {
+        // i.e. `.[foo = x] <- lst`
+        // If we find a name, we are done. The above example means "extract x from lst as foo",
+        // so we don't want to define a variable for `x` there.
+        if let Some(name) = child.child_by_field_name("name") {
+            let range = name.range();
+            let name = context.contents.node_slice(&name)?.to_string();
+            context.add_defined_variable(name.as_str(), range);
+            continue;
+        };
+
+        let Some(value) = child.child_by_field_name("value") else {
+            continue;
+        };
+
+        // i.e. `.[x, y]` where `value` is just a name that dotty assigns to
+        if value.is_identifier() {
+            let range = value.range();
+            let name = context.contents.node_slice(&value)?.to_string();
+            context.add_defined_variable(name.as_str(), range);
+            continue;
+        }
+
+        // i.e. `.[x, .[y]]` where unpacking is recursive
+        if value.is_subset() {
+            handle_assignment_dotty(&value, context)?;
+            continue;
+        }
+    }
```

Comment:
Most of my changes were adding example based docs from the perspective of someone that isn't that familiar with dotty, and also tweaking the loop structure to make it a little easier to read (i think)